Pull Request for issue markstory/asset_compress#280

elquimista opened the pull request:
I made 5 commits for solving the issue markstory/asset_compress#280. Please review the changes and merge and release the next version. Thanks.

fantastic16 committed:
Fix globbing path issue, fix ScssFilter getDependencies issue

fantastic16 committed:
Add PostCSS filter

fantastic16 committed:
Add allowDebugMode option for filters to allow some postprocessors run in debug mode

fantastic16 committed:
Filter options are merged instead of overriding in asset_compress.local.ini

fantastic16 committed:
Fix several issues relating to path expansion

markstory commented on composer.json:
```diff
@@ -1,20 +1,25 @@
 {
-  "name": "markstory/mini-asset",
+  "name": "clthck/mini-asset",
```
This needs to be changed back.

markstory commented on composer.json:
```diff
@@ -1,20 +1,25 @@
-  "keywords": ["cakephp", "assets", "minifier", "less", "coffee-script", "sass"],
-  "homepage": "https://github.com/markstory/mini-asset",
+  "keywords": ["cakephp", "assets", "minifier", "less", "coffee-script", "sass", "postcss"],
+  "homepage": "https://github.com/clthck/mini-asset",
```
Same here.

markstory commented on composer.json:
```diff
@@ -1,20 +1,25 @@
-    "issues": "https://github.com/markstory/mini-asset/issues",
-    "source": "https://github.com/markstory/mini-asset"
+    "issues": "https://github.com/clthck/mini-asset/issues",
+    "source": "https://github.com/clthck/mini-asset"
```
Same.

markstory commented on src/AssetCompiler.php:
```diff
@@ -60,16 +60,14 @@ public function __construct(FilterRegistry $filters, $debug)
-        if (!$this->debug || php_sapi_name() === 'cli') {
-            $output = $filters->output($build->path(), $output);
-        }
+        $output = $filters->output($build->path(), $output);
```
Applying minifiers all the time seems wasteful and makes debugging in the browser much harder.

elquimista replied:
This doesn't apply minifiers all the time. Inside individual postprocessor filter's `output` method, it checks whether to apply filter or not. Normally it won't apply. If `autoDebugMode` option is set to value other than zero, it will apply in debug mode as well. I thought this could be useful in some cases such as postcss' autoprefixer.
And sometimes, for debugging purpose, if we set `debug` to `false` in `app.php` and errors occur during applying postprocessors, we can't figure it out as it just says `Internal Error Occured`. In that case, we can set `debug` to `true` and check what errors and where they come from.

markstory replied:
So far output filters are minifiers, which is why they don't apply when debug is enabled.

markstory commented on src/AssetConfig.php:
```diff
@@ -180,7 +180,7 @@ public function load($path, $prefix = '')
                 // filter section.
                 $name = str_replace(self::FILTER_PREFIX, '', $section);
-                $this->filterConfig($name, $values);
+                $this->filterConfig($name, array_merge($this->filterConfig($name), $values));
```
Overwriting was intentional. With merging you must redefine any defaults that should be nulled out.

elquimista replied:
I found it might not make sense in some cases. E.g. when I'm using `PostCSS` filter, there are `options`, `path`, `node_path`, `postcss` settings. `options` should be defined in `asset_compress.ini` and the rest settings should be defined in `asset_compress.local.ini`. I think this makes more sense rather than defining all of them in either `asset_compress.ini` or local ini version.

markstory replied:
Why are all those values different? The `.local` file is generally used when people are building and deploying to different platforms/os's. In which case wouldn't all those settings need to be different?

markstory commented on src/AssetScanner.php:
```diff
@@ -90,19 +95,24 @@ protected function _expandPaths()
+     * @return void
      */
-    protected function _generateTree($path)
+    protected function _dirToArray($dir, &$result)
```
Is there a way this can be done without a pass by reference argument? From the surrounding code you could return the subtree results.

elquimista replied:
If pass by reference is needed, we have to utilize it. PHP's built in function `preg_match` uses pass by reference too.

markstory replied:
Pass by reference makes the code harder to follow and is a pattern I think should be avoided. PHP internals do all sorts of silly things that userland code generally avoids.

markstory commented on src/Filter/FilterCollection.php:
```diff
@@ -62,8 +65,12 @@ public function input($file, $content)
         foreach ($this->filters as $filter) {
-            $content = $filter->output($target, $content);
+            $settings = $filter->settings();
+            if ((array_key_exists('allowDebugMode', $settings) && !!$settings['allowDebugMode']) || $shouldApplyFilterGeneral) {
```
When are you turning on some filters in debug but not others?

elquimista replied:
`PostCSS' autoprefixer` can be applied in both debug and production mode. Others such as `uglifyjs` should be definitely applied in production mode only.

markstory replied:
PostCSS sounds like an inputFilter, not an outputFilter to me. Much less Scss and Less.

markstory commented on src/Filter/ScssFilter.php:
```diff
@@ -51,4 +54,130 @@ public function input($filename, $input)
+
+            $deps = [];
+            foreach ($imports as $name) {
+                if (preg_match('/(.*)\/\*\*\/\*$/', $name, $matches)) {
```
Having some tests to ensure this behavior continues to work would be nice. I've never seen an import with `**` in the name.

elquimista replied:
To be honest, I don't have experience in writing or using `phpunit` at all. I agree that we need tests for this, but I'm afraid I am not skillful enough to do that at the moment.
If you google about `sass_globbing` (it's a ruby gem), it uses that pattern for matching all the files recursively inside specified path.

markstory replied:
I took a look, and I don't think it will be viable to support every sass extension that exists, so I'd rather not add support for sass_glob.

markstory commented on src/Filter/ScssFilter.php:
```diff
@@ -51,4 +54,130 @@ public function input($filename, $input)
+            $basepath = dirname($basepath);
+        }
+        array_unshift($paths, $basepath);
+        return $paths;
```
~~This looks like duplicate code from the AssetScanner.~~ Expanding the `**` operator seems like something that should happen during AssetScanner's path expansion and not inside a specific filter.

elquimista replied:
We surely need expanding the *\* operator in AssetScanner's path expansion. And we need the same for ScssFilter as well. That's because ScssFilter needs to be aware of its dependency scss files. Getting dependencies vary from filter to filter, doesn't it? To my knowledge, AssetScanner's path expansion is needed for locating build file paths.

markstory commented on src/Filter/Sprockets.php:
```diff
@@ -139,21 +139,23 @@ public function getDependencies(AssetTarget $target)
+                    $path = $this->_findFile($matches[2][$i]);
+                }
+                $dep = new Local($path);
+                $children[] = $dep;
```
What issue is this solving? Could tests be added?

elquimista replied:
Let's suppose we have the following lines in `application.js`.

``` js
//= require <jquery.min.js>
//= require <underscore-min.js>
//= require "base/base.js"
```

I found that application.js did not update when underscore-min.js or base.js files were modified. But it did update when jquery.min.js file was changed.
`preg_match` in the original code returns only the first match. We need global matching, so I thought `preg_match` should be replaced with `preg_match_all` and had to write some code to handle array of lines accordingly.

markstory replied:
Ok, I'll take a look at that and try and reproduce in the test suite.

markstory replied:
Thanks for the info on how to reproduce the problem . I think I've addressed this issue and the scss dependency issue in #19.

markstory commented on src/Filter/ScssFilter.php:
```diff
@@ -51,4 +54,130 @@ public function input($filename, $input)
+                $pathinfo = pathinfo($name);
+                if (substr($pathinfo['basename'], 0, 1) !== '_') {
+                    $name = $pathinfo['dirname'] . '/_' . $pathinfo['basename'];
+                }
```
This looks like a change that will break existing applications that do not have `_` prefixes.

elquimista replied:
By convention, all imported file names should begin with underscore (_) in sass.

Commit:
Update composer.json

Commit:
Fix getDependencies in ScssFilter.php

Allow underscored and non-underscored version of filenames in @import, in `getDependencies` method.

elquimista commented:
So you are not going to add sass_globbing feature support?

Commit:
Update composer.json again

Commit:
Change repo name and update composer.json accordingly

markstory commented:
> So you are not going to add sass_globbing feature support?

From my understanding this is not a feature of sass, but an optional module that can be added on top. I'm not sure that can of worms is worth opening.
